File: Python/mainConvHullPBDTwoChain.py
import pyPBAD

from utils import *
from visualizer import *

if __name__=='__main__':
    floor=pyPBAD.BBoxExact([-150,-150,-8],[30,30,-6])

    bodies=[]
    #first chain
    bodies.append(
        pyPBAD.ArticulatedLoader.createChain(int(pyPBAD.JOINT_TYPE.TRANS_3D) | int(pyPBAD.JOINT_TYPE.ROT_3D_EXP), 20, 0.5, 0.24,
                                          D2R(30), 0, 0, 0, 3, 0, 0, 0))
    #second chain
    bodies.append(
        pyPBAD.ArticulatedLoader.createChain(int(pyPBAD.JOINT_TYPE.TRANS_3D) | int(pyPBAD.JOINT_TYPE.ROT_3D_EXP), 20, 0.5, 0.24,
                                          D2R(30), 0, 0, 0, 3, 0, 0, 0))
    #third chain
    bodies.append(
        pyPBAD.ArticulatedLoader.createChain(int(pyPBAD.JOINT_TYPE.TRANS_3D) | int(pyPBAD.JOINT_TYPE.ROT_3D_EXP), 20, 0.5, 0.24,
                                          D2R(30), 0, 0, 0, 3, 0, 0, 0))
    #forth chain
    bodies.append(
        pyPBAD.ArticulatedLoader.createChain(int(pyPBAD.JOINT_TYPE.ROT_3D_EXP), 20, 0.5, 0.24,
                                          D2R(30), 0, 0, 0, 3, 0, 0, 0))
    #combine
    body=pyPBAD.ArticulatedBody()
    pyPBAD.ArticulatedUtils(body).combine(bodies)
    #offset
    trans=getBodyJointTrans(body,22)
    trans[2,3]=1
    setBodyJointTrans(body,22,trans)
    #offset
    trans=getBodyJointTrans(body,43)
    trans[2,3]=2
    setBodyJointTrans(body,43,trans)
    #offset
    trans=getBodyJointTrans(body,64)
    trans[2,3]=3
    setBodyJointTrans(body,64,trans)
    
    #simulator
    sim=pyPBAD.ConvHullPBDSimulator(0.033)
    sim.setArticulatedBody(body)
    sim.addShape(floor)
    sim.gravity=[0,0,-9.81]
    sim.setOutput(True)

    #setup kinematic 1
    def PFunc1(time,d):
        return np.array([math.cos(time*2),math.sin(time*2),0])*4.5
    def DFunc1(time,d):
        return np.array([-math.sin(time*2),math.cos(time*2),0])*9
    for k in range(1,3):
        #control
        ctrl=getBodyJointControl(body,k)
        for d in range(ctrl.shape[0]):
            ctrl[d]=1
        setBodyJointControl(body,k,ctrl)
        print(f"Joints[{k}] has {body.getJoint(k).nrDOF()} DOF")
        #param
        param=sim.getJointPhysicsParameter(k)
        param.kp=1000
        param.kd=10
        param.setTarPFunc(PFunc1)
        param.setTarDFunc(DFunc1)
        sim.setJointPhysicsParameter(k,param)

    #setup kinematic 2
    def PFunc2(time,d):
        return np.array([math.cos(time*2),math.sin(time*2),0])*3
    def DFunc2(time,d):
        return np.array([-math.sin(time*2),math.cos(time*2),0])*6
    for k in range(22,24):
        #control
        ctrl=getBodyJointControl(body,k)
        for d in range(ctrl.shape[0]):
            ctrl[d]=1
        setBodyJointControl(body,k,ctrl)
        print(f"Joints[{k}] has {body.getJoint(k).nrDOF()} DOF")
        #param
        param=sim.getJointPhysicsParameter(k)
        param.kp=1000
        param.kd=10
        param.setTarPFunc(PFunc2)
        param.setTarDFunc(DFunc2)
        sim.setJointPhysicsParameter(k,param)

    #setup kinematic 3
    def PFunc3(time,d):
        return np.array([math.cos(time*2),math.sin(time*2),0])*1.5
    def DFunc3(time,d):
        return np.array([-math.sin(time*2),math.cos(time*2),0])*3
    for k in range(43,45):
        #control
        ctrl=getBodyJointControl(body,k)
        for d in range(ctrl.shape[0]):
            ctrl[d]=1
        setBodyJointControl(body,k,ctrl)
        print(f"Joints[{k}] has {body.getJoint(k).nrDOF()} DOF")
        #param
        param=sim.getJointPhysicsParameter(k)
        param.kp=1000
        param.kd=10
        param.setTarPFunc(PFunc3)
        param.setTarDFunc(DFunc3)
        sim.setJointPhysicsParameter(k,param)

    #run simulator
    runFrames=False
    if runFrames:
        frames=[]
        for i in range(10):
            print(f"Computing frame {i}")
            sim.step()
            frames.append(sim.pos)
    else: frames=None

    #visualize
    vis = Visualizer(sim, frames)
    vis.mainLoop()

File: Main/DebugSimulator/mainConvHullPBDTwoChain.cpp
#include <Articulated/ArticulatedUtils.h>
#include <Articulated/ArticulatedLoader.h>
#include <Simulator/SimulatorVisualizer.h>
#include <Simulator/ConvHullPBDSimulator.h>
#include <Simulator/MeshBasedPBDSimulator.h>
#include <chrono>

using namespace PHYSICSMOTION;

int main(int argc,char** argv) {
  typedef FLOAT T;
  DECL_MAT_VEC_MAP_TYPES_T
  //create body
  std::vector<ArticulatedBody> bodies(2);
  std::shared_ptr<ArticulatedBody> body(new ArticulatedBody);
  std::shared_ptr<ShapeExact> floor(new BBoxExact(BBoxExact::Vec3T(-150,-150,-8),BBoxExact::Vec3T(30,30,-1)));
  {
    tinyxml2::XMLDocument pt;
    pt.InsertEndChild(pt.NewElement("root"));
    ArticulatedLoader::createChain(*(pt.RootElement()),Joint::TRANS_3D|Joint::ROT_3D_EXP,5,.5f,0.1f,D2R(60),0,0,0,3,0,0,0);
    ArticulatedUtils utils2(bodies[0]);
    utils2.assemble(*(pt.RootElement()));
  }
  {
    tinyxml2::XMLDocument pt;
    pt.InsertEndChild(pt.NewElement("root"));
    ArticulatedLoader::createChain(*(pt.RootElement()),Joint::ROT_3D_EXP,5,.5f,0.1f,D2R(60),0,0,0,3,0,0,0);
    ArticulatedUtils utils2(bodies[1]);
    utils2.assemble(*(pt.RootElement()));
  }
  ArticulatedUtils(*body).combine(bodies);
  CTR(body->joint(7)._trans)=Vec3T(0,0,.5).template cast<ArticulatedBody::T>();
  //simulator
  //MeshBasedPBDSimulator sim(0.01f);
  ConvHullPBDSimulator sim(0.01f);
  sim.setArticulatedBody(body);
  sim.addShape(floor);
  sim.setGravity(Vec3T(0,0,-9.81f));
  sim.setOutput(true);
  //visualizer
  SimulatorVisualizer render(sim);
  render.setLightSize(0);
  std::vector<Eigen::Matrix<float,3,1>>ArticulateDiffuse;
  Eigen::Matrix<float,3,1> LightDiffuse=Eigen::Matrix<float,3,1>(.5,.5,.5);
  render.setLightDiffuse(LightDiffuse);
  //setup kinematic
  /*for(int k=1; k<2; k++) {
    //control
    Joint& J=body->joint(k);
    J._control.setOnes(J.nrDOF());
    std::cout<<k<<" "<<J.nrDOF()<<std::endl;
    //param
    auto& param=sim.getJointPhysicsParameter(k);
    param._kp=1e3;
    param._kd=1e1;
    param._tarP=[&](T time,int)->Vec {
      return Vec3T(cos(time*2),sin(time*2),0)*4.5;
    };
    param._tarD=[&](T time,int)->Vec {
      return Vec3T(-sin(time*2),cos(time*2),0)*9;
    };
  }
  for(int k=12; k<13; k++) {
    //control
    Joint& J=body->joint(k);
    J._control.setOnes(J.nrDOF());
    std::cout<<k<<" "<<J.nrDOF()<<std::endl;
    //param
    auto& param=sim.getJointPhysicsParameter(k);
    param._kp=1e3;
    param._kd=1e1;
    param._tarP=[&](T time,int)->Vec {
      return Vec3T(cos(time*2),sin(time*2),0)*3;
    };
    param._tarD=[&](T time,int)->Vec {
      return Vec3T(-sin(time*2),cos(time*2),0)*6;
    };
  }*/
  visualizeSimulator(argc,argv,sim);
  //run app
  auto start = std::chrono::high_resolution_clock::now();
  std::vector<Vec> traj;
  traj.resize(100);
  for(int i=0;i<100;i++) {
    std::cout<<i<<std::endl;
    sim.step();
    traj.push_back(sim.pos());
  }
  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed_time = end - start;
  std::cout << "Elapsed time: " << elapsed_time.count() << " seconds" << std::endl;
  int frame=0;
  render.visualize(&traj,&frame);
  return 0;
}
